packages/wonder-blocks-modal/src/components/drawer-launcher.tsx

```diff
@@ -0,0 +1,549 @@
+     * themselves, since this will capture a more complete set of close events.
+     *
+     * Called when the modal needs to notify the parent component that it should
+     * be closed.
+     *
+     * This prop must be used when the component is being used as a controlled
+     * component.
+     */
+    onClose?: () => unknown;
+
+    /**
+     * WARNING: This props should only be used when using the component as a
+     * controlled component.
+     */
+    children?: (arg1: {
+        openModal: () => unknown;
+        styles?: FlexibleDialogStyles;
+    }) => React.ReactNode;
+}> &
+    WithActionSchedulerProps;
+
+// Set a default timing duration for animations and focus management. Also used for tests.
+export const DEFAULT_TIMING_DURATION = 400;
+
+const defaultProps = {
+    backdropDismissEnabled: true,
+    defaultTimingDuration: DEFAULT_TIMING_DURATION,
+} as const;
+
+const DrawerLauncher = (props: Props) => {
+    const {
+        modal,
+        backdropDismissEnabled = defaultProps.backdropDismissEnabled,
+        initialFocusId,
+        closedFocusId,
+        testId,
+        opened: controlledOpened,
+        onClose,
+        children,
+        schedule,
+        alignment,
+        styles,
+        animated = true,
+        timingDuration = defaultProps.defaultTimingDuration,
+    } = props;
+
+    // State for uncontrolled mode
+    const [uncontrolledOpened, setUncontrolledOpened] = React.useState(false);
+    // State to track exit animation
+    const [isExiting, setIsExiting] = React.useState(false);
+
+    const launcherRef = React.useRef<HTMLDivElement>(null);
+
+    // Get Language direction from closest parent with dir attribute
+    const [direction, setDirection] = React.useState<string | undefined>(
+        undefined,
+    );
+
+    // Ref to store the last focused element
+    const lastElementFocusedOutsideModalRef = React.useRef<HTMLElement | null>(
+        null,
+    );
+
+    // Determine if we're in controlled or uncontrolled mode
+    const opened =
+        typeof controlledOpened === "boolean"
+            ? controlledOpened
+            : uncontrolledOpened;
+
+    // Validation warnings
+    React.useEffect(() => {
+        if (process.env.NODE_ENV !== "production") {
+            if (typeof controlledOpened === "boolean" && children) {
+                // eslint-disable-next-line no-console
+                console.warn("'children' and 'opened' can't be used together");
+            }
+            if (typeof controlledOpened === "boolean" && !onClose) {
+                // eslint-disable-next-line no-console
+                console.warn("'onClose' should be used with 'opened'");
+            }
+            if (typeof controlledOpened !== "boolean" && !children) {
+                // eslint-disable-next-line no-console
+                console.warn("either 'children' or 'opened' must be set");
+            }
+        }
+    }, [controlledOpened, children, onClose]);
```

Comment:
Do you think we could structure the prop types to make sure invalid prop combinations can never be used together? I haven't seen runtime prop validation in WB components, and I'm not sure if logs are monitored by teams during development so these messages might be missed! It would be nice if we can avoid these prop combinations earlier during build time with typechecks!

It is nice though how clear these warning messages are, compared to TS errors 😅 What do you think?

Reply:
I totally agree that these messages would probably get missed. TS errors would be a lot more appropriate! I did copy them over from ModalLauncher, but I'll see if I can get some feedback from typing instead. I have a similar thing happening with validation for the new `DrawerDialog` to ensure it is the only one used with `DrawerLauncher`. Runtime messages aren't a good fit for development workflows!

Reply:
Removed these runtime warnings in favor of stronger typing!